lib/modules.nix

```diff
@@ -412,10 +413,43 @@ rec {
       # modules recursively. It returns the final list of unique-by-key modules
       filterModules = modulesPath: { disabled, modules }:
         let
-          moduleKey = m: if isString m && (builtins.substring 0 1 m != "/")
-            then toString modulesPath + "/" + m
-            else toString m;
-          disabledKeys = map moduleKey disabled;
+          moduleKey = file: m:
+            if isString m && (builtins.substring 0 1 m != "/")
+            then
+              toString modulesPath + "/" + m
+
+            else if isPath m || isString m
+            then toString m
+
+            else if m?key
+            then
+              # anonymous module with key
+              m.key
+
+            else if isFunction m
+            then
+              let mAttrs = m fakeArgs;
+                  fakeArgs =
+                    mapAttrs
+                      (k: v: throw "If a module in 'disabledModules' is not referenced by its path, it must be identified by a 'key' attribute. Retrieving this attribute failed for ${theModule} in the 'disabledModules' of a module in '${file}', because its 'key' attribute requires the module argument '${lib.strings.escapeNixIdentifier k}' to be computed first. This is not possible, because the set of applicable modules must be determined before any configuration can be computed.")
```

Comment:
I don't think this should be part of it. Flake-aware code can map over `nixosModules.*`, wrapping each using `module: { key = "<flake>#<attr>"; imports = [ module ]; }`. Providing fake arguments won't be necessary like that.

Also by not allowing functions, this change can be seen as just allowing modules to be coerced to a string in a different way using `.key`, similar to `.outPath` and `.__toString`, which sounds more reasonable to me, because evidently from this code here, `disabledModules` and `imports` are fundamentally different (`disabledModules` only needs a reference to a module, while `imports` needs the module itself).

However an alternative idea I'd be more in favor of is to require all `imports` to have a `key`. So you couldn't just do `imports = [ {} ]` anymore, it would have to be `imports = [ { key = "something"; } ]`, the key can't be behind the fixed point though. This then ensures that all modules have a stable reference that can be used to disable them. And it would also allow using all modules in both `imports` and `disabledModules`.

Reply:
I've removed the possibility to specify a key in a function. It's good to be careful like that, and I don't think it's a problem for users to remove this functionality.

> However an alternative idea I'd be more in favor of is to require all `imports` to have a `key`.

I don't see the benefit. The benefit comes from having more keys. Requiring those is a rather harsh incentive for such a small improvement.

Anyway, that's just future work if anything.